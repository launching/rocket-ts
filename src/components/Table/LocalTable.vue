<script lang="ts">
import {
  Component,
  Prop,
  Vue,
  PropSync,
  Inject,
  Watch,
  Model,
  Emit,
  Mixins,
} from 'vue-property-decorator';
import { Column, FormModel } from '@/components/Interface';
import BaseTable from './BaseTable.vue';
import RForm from '@/components/Form/Form.vue';
import RTableColumn from '@/components/Table/TableColumn.vue';
import _ from 'lodash';
@Component({
  components: {
    RForm,
    RTableColumn,
  },
})
export default class RLocalTable extends Mixins(BaseTable) {
  search(params: FormModel, page: number = 0) {
    this.filterData = {
      ...params,
    };
  }

  originData: Array<any> = [];

  async dataHandler() {
    return _.isFunction(this.data) ? this.data() : this.data;
  }

  @Watch('data', { immediate: true, deep: true })
  async refreshData() {
    this.loading = true;
    const res = await this.dataHandler();
    this.loading = false;
    if (!_.isArray(res)) {
      throw new Error('data is must be Array type');
    }
    this.originData = res;
  }

  get totalTable() {
    return this.originData.filter(item => {
      let temp = true;
      for (var key in this.filterData) {
        if (!this.filterData[key] && this.filterData[key] !== false) {
          continue;
        }

        if (_.isString(item[key])) {
          temp = temp && !!~(item[key] as string).indexOf(this.filterData[key]);
        } else {
          temp = temp && +item[key] === +this.filterData[key];
        }
      }
      return !!temp;
    });
  }

  get targetData() {
    const { page, pageSize } = this.pagination;
    return this.totalTable.slice((page - 1) * pageSize, pageSize * page);
  }

  get total() {
    return this.totalTable.length;
  }
}
</script>
<style lang="stylus" scoped>
.r-table
  .pagination-container
    padding 32px 16px
</style>
