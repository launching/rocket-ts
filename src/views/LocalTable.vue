<template lang="pug">
r-local-table(:filter="filter" :columns="columns" :data="data" border fit hightlight-current-row :localPage="true" ref="table")
</template>
<script>
import RLocalTable from '@/components/Table/LocalTable.vue';
import * as api from '@/api/articles';

export default {
  components: {
    RLocalTable,
  },
  data() {
    return {
      filter: {
        inline: true,
        children: [
          { widget: 'input', name: 'title' },
          {
            widget: 'select',
            name: 'importance',
            clearable: true,
            options: [
              {
                label: 1,
                value: 1,
              },
              {
                label: 2,
                value: 2,
              },
              {
                label: 3,
                value: 3,
              },
            ],
          },
          {
            widget: 'select',
            name: 'type',
            clearable: true,
            options: [
              { value: 'CN', label: 'China' },
              { value: 'US', label: 'USA' },
              { value: 'JP', label: 'Japan' },
              { value: 'EU', label: 'Eurozone' },
            ],
          },
        ],
        toolbar: [
          {
            icon: 'el-icon-search',
            type: 'primary',
            text: '搜索',
            action: 'SEARCH',
          },
          {
            text: '取消',
            action: 'CANCEL',
          },
          { icon: 'el-icon-edit', text: '添加', action() {} },
          { icon: 'el-icon-download', text: '导出', action() {} },
        ],
      },
      border: true,
      hightlightCurrentRow: true,
      fit: true,
      columns: [
        { label: '序号', name: 'id', width: '80' },
        { label: '时间', name: 'timestamp' },
        { label: '标题', name: 'title' },
        { label: '作者', name: 'author', width: '180' },
        { label: '重要性', name: 'importance', width: '110' },
        { label: '阅读数', name: 'pageviews', width: '95' },
        { label: '状态', name: 'status', type: 'tag', width: '100' },
        {
          label: '操作',
          type: 'toolbar',
          width: '230',
          children: [
            {
              type: 'primary',
              text: '编辑',
              size: 'mini',
            },
            {
              text: '草稿',
              size: 'mini',
              premise: child => child.status !== 'draft',
              action: child => this.handleModifyStatus(child, 'draft'),
            },
            {
              text: '删除',
              size: 'mini',
              type: 'danger',
              premise: child => child.status !== 'deleted',
              action: child => this.handleModifyStatus(child, 'deleted'),
            },
            {
              text: '发布',
              size: 'mini',
              type: 'success',
              premise: child => child.status !== 'published',
              action: child => this.handleModifyStatus(child, 'published'),
            },
          ],
        },
      ],
      data() {
        return api.getArticles().then(res => res.data.items);
      },
    };
  },
  methods: {
    handleModifyStatus(row, status) {
      this.$message({
        message: '操作成功',
        type: 'success',
      });
      row.status = status;
    },
  },
  created() {},
};
</script>
