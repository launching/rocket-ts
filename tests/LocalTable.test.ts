import Vue from 'vue';
import ElementUI from 'element-ui';
Vue.use(ElementUI);

import LocalTable from '@/components/Table/LocalTable.vue';

import { shallowMount, mount } from '@vue/test-utils';

describe('LocalTable.vue', () => {
  const columns = [
    { label: '序号', name: 'id', width: '80' },
    { label: '时间', name: 'timestamp', width: '180' },
    { label: '标题', name: 'title', width: '150' },
    { label: '作者', name: 'author', width: '180' },
    { label: '重要性', name: 'importance', width: '110' },
    { label: '阅读数', name: 'pageviews', width: '95' },
    { label: '状态', name: 'status', type: 'tag', width: '100' },
    {
      label: '操作',
      type: 'toolbar',
      width: '230',
      children: [
        {
          type: 'primary',
          text: '编辑',
          size: 'mini',
        },
        {
          text: '草稿',
          size: 'mini',
        },
        {
          text: '删除',
          size: 'mini',
          type: 'danger',
        },
        {
          text: '发布',
          size: 'mini',
          type: 'success',
        },
      ],
    },
  ];

  const filter = {
    inline: true,
    children: [
      { widget: 'input', name: 'title' },
      {
        widget: 'select',
        name: 'importance',
        clearable: true,
        options: [
          {
            label: 1,
            value: 1,
          },
          {
            label: 2,
            value: 2,
          },
          {
            label: 3,
            value: 3,
          },
        ],
      },
      {
        widget: 'select',
        name: 'type',
        clearable: true,
        options: [
          { value: 'CN', label: 'China' },
          { value: 'US', label: 'USA' },
          { value: 'JP', label: 'Japan' },
          { value: 'EU', label: 'Eurozone' },
        ],
      },
    ],
    toolbar: [
      {
        icon: 'el-icon-search',
        type: 'primary',
        text: '搜索',
      },
      { icon: 'el-icon-edit', text: '添加', action() {} },
      { icon: 'el-icon-download', text: '导出', action() {} },
    ],
  };

  const hightlightCurrentRow = true;

  const fit = true;

  const data: any = [];
  const border = true;

  it('localTable Snapshot', async () => {
    const wrapper = shallowMount(LocalTable, {
      propsData: {
        columns,
        filter,
        data,
        border,
        fit,
        hightlightCurrentRow,
      },
    });

    await wrapper.vm.$nextTick();
    // expect(wrapper.html()).toMatchSnapshot();
  });

  it('data is function', callback => {
    const wrapper = shallowMount(LocalTable, {
      propsData: {
        columns,
        filter,
        data() {
          return [
            {
              id: '1',
              timestamp: '212',
              title: '123',
              author: '12312',
              importance: '1232',
              pageviews: '124',
              status: '123',
            },
            {
              id: '2',
              timestamp: '212',
              title: '123',
              author: '12312',
              importance: '1232',
              pageviews: '124',
              status: '123',
            },
          ];
        },
        border,
        fit,
        hightlightCurrentRow,
      },
    });

    setTimeout(() => {
      // expect(wrapper.html()).toMatchSnapshot();
      let vm: any = wrapper.vm;
      expect(vm.targetData.length).toBe(2);
      callback();
    }, 200);
  });
});
