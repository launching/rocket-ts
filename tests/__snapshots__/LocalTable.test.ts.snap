// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalTable.vue data is function 1`] = `
<div class="r-table">
  <r-form-stub children="[object Object],[object Object],[object Object]" defaultmodel="[object Object]" toolbar="[object Object],[object Object],[object Object]" inline="true"></r-form-stub>
  <el-table-stub data="[object Object],[object Object]" fit="true" border="true" showheader="true" selectonindeterminate="true" indent="16" treeprops="[object Object]" hightlight-current-row="true" class="">
    <r-table-column-stub label="序号" width="80" name="id"></r-table-column-stub>
    <r-table-column-stub label="时间" width="180" name="timestamp"></r-table-column-stub>
    <r-table-column-stub label="标题" width="150" name="title"></r-table-column-stub>
    <r-table-column-stub label="作者" width="180" name="author"></r-table-column-stub>
    <r-table-column-stub label="重要性" width="110" name="importance"></r-table-column-stub>
    <r-table-column-stub label="阅读数" width="95" name="pageviews"></r-table-column-stub>
    <r-table-column-stub label="状态" width="100" name="status" type="tag"></r-table-column-stub>
    <r-table-column-stub label="操作" width="230" type="toolbar" children="[object Object],[object Object],[object Object],[object Object]"></r-table-column-stub>
    <div class="el-loading-mask" style="display: none;">
      <div class="el-loading-spinner"><svg viewBox="25 25 50 50" class="circular">
          <circle cx="50" cy="50" r="20" fill="none" class="path"></circle>
        </svg>
        <!---->
      </div>
    </div>
  </el-table-stub>
  <div class="pagination-container">
    <el-pagination-stub pagesize="10" total="2" pagercount="7" currentpage="1" layout="total, sizes, prev, pager, next, jumper" pagesizes="10,20,30,40" background="true"></el-pagination-stub>
  </div>
</div>
`;

exports[`LocalTable.vue localTable Snapshot 1`] = `
<div class="r-table">
  <r-form-stub children="[object Object],[object Object],[object Object]" defaultmodel="[object Object]" toolbar="[object Object],[object Object],[object Object]" inline="true"></r-form-stub>
  <el-table-stub data="" fit="true" border="true" showheader="true" selectonindeterminate="true" indent="16" treeprops="[object Object]" hightlight-current-row="true" class="el-loading-parent--relative">
    <r-table-column-stub label="序号" width="80" name="id"></r-table-column-stub>
    <r-table-column-stub label="时间" width="180" name="timestamp"></r-table-column-stub>
    <r-table-column-stub label="标题" width="150" name="title"></r-table-column-stub>
    <r-table-column-stub label="作者" width="180" name="author"></r-table-column-stub>
    <r-table-column-stub label="重要性" width="110" name="importance"></r-table-column-stub>
    <r-table-column-stub label="阅读数" width="95" name="pageviews"></r-table-column-stub>
    <r-table-column-stub label="状态" width="100" name="status" type="tag"></r-table-column-stub>
    <r-table-column-stub label="操作" width="230" type="toolbar" children="[object Object],[object Object],[object Object],[object Object]"></r-table-column-stub>
    <div class="el-loading-mask el-loading-fade-leave el-loading-fade-leave-active" style="">
      <div class="el-loading-spinner"><svg viewBox="25 25 50 50" class="circular">
          <circle cx="50" cy="50" r="20" fill="none" class="path"></circle>
        </svg>
        <!---->
      </div>
    </div>
  </el-table-stub>
  <div class="pagination-container">
    <el-pagination-stub pagesize="10" total="0" pagercount="7" currentpage="1" layout="total, sizes, prev, pager, next, jumper" pagesizes="10,20,30,40" background="true"></el-pagination-stub>
  </div>
</div>
`;
